cadc-util/src/main/java/org/opencadc/persist/Entity.java

```diff
@@ -307,7 +307,84 @@ public URI computeMetaChecksum(MessageDigest digest) {
                 + this.getClass().getName(), e);
         }
     }
-    
+
+    /**
+     * Visit the state of the entity in the standard traversal order: depth first, alphabetic.
+     * @param ev 
+     */
+    public void visit(EntityVisitor ev) {
+        visitImpl(this.getClass(), this, ev);
+        
+    }
+
+    private void visitImpl(Class c, Object o, EntityVisitor ev) {
+        try { 
+            if (o instanceof Entity) {
+                Entity ce = (Entity) o;
+                ev.visitLeaf("Entity.id", o);
+                ev.visitLeaf("Entity.metaProducer", ce.metaProducer);
+            }
+            
+            SortedSet<Field> fields = getStateFields(c);
+            for (Field f : fields) {
+                String cf = f.getDeclaringClass().getSimpleName() + "." + f.getName();
+                f.setAccessible(true);
+                Object fo = f.get(o);
+                
+                
+                if (fo == null) {
+                    ev.visitNull(cf);
+                } else {
+                    Class ac = fo.getClass();
+                    if (ac.isEnum() || PrimitiveWrapper.class.isAssignableFrom(ac)) {
+                        try {
+                            log.warn("unwrap: " + ac.getSimpleName() + ".getValue()");
+                            Method m = ac.getMethod("getValue");
+                            Object val = m.invoke(fo);
+                            ev.visitLeaf(cf, val);
+                        } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException ex) {
+                            throw new RuntimeException("BUG - enum " + ac.getName() + " does not have getValue()", ex);
+                        }
+                    } else if (isDataModelClass(ac)) {
+                        // depth-first recursion
+                        visitImpl(ac, fo, ev);
+                        // visit intermediate DM class
+                        ev.visitNode(cf, fo);
+                    } else if (fo instanceof Collection) {
+                        Collection stuff = (Collection) fo;
+                        if (!stuff.isEmpty()) {
+                            Iterator i = stuff.iterator();
+                            while (i.hasNext()) {
+                                Object co = i.next();
+                                Class cc = co.getClass();
```

Comment:
Isn't the next code identical with L338 and down? Can it be factored out in a method?

Reply:
It turns out to be more complex than I liked because the trailing `} else {` is slightly different inside the _is-collection_ block... I found the repetition in the same method more clear than trying to call a method.